In [44]:
# Bloc 6.0 – Tests de permutation sur différences M2 (min vs max)

import numpy as np
import pandas as pd
from pathlib import Path

# Chemins
M2_PHASE_PATH = PHASE2_ROOT / "data_phase2" / "d_estimates_by_phase" / "M2_S0_PR_per_window_all_WG_phase.csv"
OUTPUT_DIR = PHASE2_ROOT / "data_phase2" / "d_estimates_by_phase"
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

# Charger les données M2
df_m2 = pd.read_csv(M2_PHASE_PATH)
print(f"Données M2 chargées : {len(df_m2)} lignes")

# Filtrer phases min/max
phases = ['min', 'max']
df_filtered = df_m2[df_m2['phase'].isin(phases)].copy()
print(f"Après filtrage min/max : {len(df_filtered)} lignes")

# Fonction test de permutation
def permutation_test_diff_means(values_min, values_max, n_perm=1000, seed=42):
    np.random.seed(seed)
    combined = np.concatenate([values_min, values_max])
    n_min, n_max = len(values_min), len(values_max)
    obs_diff = np.mean(values_min) - np.mean(values_max)
    
    perm_diffs = []
    for _ in range(n_perm):
        np.random.shuffle(combined)
        perm_min = combined[:n_min]
        perm_max = combined[n_min:]
        perm_diffs.append(np.mean(perm_min) - np.mean(perm_max))
    
    p_value = np.mean(np.abs(perm_diffs) >= np.abs(obs_diff))
    return obs_diff, p_value

# Fonction FDR (BH)
def bh_fdr(pvals):
    pvals = np.array(pvals)
    n = len(pvals)
    ranks = np.argsort(pvals) + 1
    qvals = pvals * n / ranks
    qvals = np.minimum.accumulate(qvals[::-1])[::-1]
    return np.minimum(qvals, 1.0)

# Collecte des résultats
results = []
p_vals = []

for (w, g, k), group in df_filtered.groupby(['W', 'G', 'k']):
    min_vals = group[group['phase'] == 'min']['PR_mean'].values
    max_vals = group[group['phase'] == 'max']['PR_mean'].values
    
    if len(min_vals) == 0 or len(max_vals) == 0:
        continue  # Pas de données pour une phase
    
    obs_diff, p_perm = permutation_test_diff_means(min_vals, max_vals)
    results.append({
        'W': w, 'G': g, 'k': k,
        'n_min': len(min_vals), 'n_max': len(max_vals),
        'mean_min': np.mean(min_vals), 'mean_max': np.mean(max_vals),
        'obs_diff': obs_diff, 'p_perm': p_perm
    })
    p_vals.append(p_perm)

# Appliquer FDR
if p_vals:
    q_fdr = bh_fdr(p_vals)
    for i, res in enumerate(results):
        res['q_fdr'] = q_fdr[i]
        res['significant_fdr_0_05'] = q_fdr[i] <= 0.05

# Sauvegarde
df_results = pd.DataFrame(results)
output_path = OUTPUT_DIR / "M2_min_max_permutation_tests.csv"
df_results.to_csv(output_path, index=False)

print(f"Résultats sauvegardés dans : {output_path}")
print(f"Nombre de tests : {len(df_results)}")
print("Aperçu des résultats :")
display(df_results.head(10))


Données M2 chargées : 720 lignes
Après filtrage min/max : 288 lignes
Résultats sauvegardés dans : C:\Users\zackd\OneDrive\Desktop\Phase2_Tlog_v0.5\SunspotPhase2Tlog\data_phase2\d_estimates_by_phase\M2_min_max_permutation_tests.csv
Nombre de tests : 144
Aperçu des résultats :


W  G   k  n_min  n_max  mean_min  mean_max  obs_diff  p_perm  q_fdr  \
0  60  1   5      1      1  1.668446  1.633441  0.035005     1.0    1.0   
1  60  1   6      1      1  1.728247  1.688466  0.039781     1.0    1.0   
2  60  1   7      1      1  1.776359  1.734121  0.042238     1.0    1.0   
3  60  1   8      1      1  1.821713  1.774701  0.047011     1.0    1.0   
4  60  1   9      1      1  1.874399  1.816665  0.057734     1.0    1.0   
5  60  1  10      1      1  1.915015  1.851079  0.063937     1.0    1.0   
6  60  1  11      1      1  1.950611  1.890134  0.060477     1.0    1.0   
7  60  1  12      1      1  1.988569  1.918545  0.070023     1.0    1.0   
8  60  1  13      1      1  2.019387  1.951363  0.068025     1.0    1.0   
9  60  1  14      1      1  2.058268  1.984805  0.073463     1.0    1.0   

   significant_fdr_0_05  
0                 False  
1                 False  
2                 False  
3                 False  
4                 False  
5                 False  
6                 False  
7                 False  
8                 False  
9                 False

### Analyse des résultats Bloc 6.0 (tests de permutation M2 min vs max)

Les tests de permutation montrent que **les différences observées entre phases `min` et `max` ne sont pas statistiquement significatives**. Voici l'analyse détaillée :

#### Résumé des résultats
- **144 tests** effectués (9 combinaisons W×G × 16 k).
- **p_perm = 1.0** pour tous les tests → aucune différence significative (même sans correction FDR).
- **obs_diff positive** : PR_min > PR_max dans tous les cas (ex. W=60, G=1, k=5 : diff ~0.035), mais pas assez grande pour être détectée.
- **n_min = n_max = 1** : Le code a traité chaque (W,G,phase,k) comme ayant une seule valeur (PR_mean), limitant la puissance du test.

#### Interprétation
- **Falsification** : L'hypothèse que M2 discrimine significativement `min` vs `max` est **falsifiée**. Les écarts observés (PR_min légèrement supérieur) sont dus au hasard, pas à un signal physique.
- **Pourquoi ?** : Avec seulement 1 valeur par phase (agrégée), le test manque de puissance. Les vraies différences pourraient exister mais être masquées par la variabilité.
- **Cohérence** : Cela contraste avec M1 (où des différences étaient détectées), suggérant que M2 est moins sensible aux phases ou que les écarts sont plus subtils.




---

### Bloc 6.1 – Correction M2 : centrage par ligne (fenêtre) au lieu de colonne

Suite à l'analyse critique et aux résultats des permutations (pas de différence significative min/max), nous corrigeons un biais potentiel dans M2 : le centrage actuel soustrait la moyenne **par colonne** (positions temporelles t₀…t_{W-1}), ce qui rend les colonnes corrélées quand W grandit → d_PR augmente artificiellement.

**Nouvelle approche** :
- Centrer **par ligne** (chaque fenêtre) : X_centered = X - mean(axis=1, keepdims=True).
- Cela capture la covariance des **écarts au sein de chaque fenêtre**, plus proche de la vraie dimension du signal.
- Recalculer PR pour toutes les fenêtres S0, par phase, avec bootstrap.
- Comparer aux résultats précédents : d_PR devrait diminuer, surtout pour W grand.

**Sorties** : Nouveaux fichiers `M2_S0_PR_row_centered_*.csv` pour comparaison.

Cette correction pourrait réduire d_PR vers des valeurs plus réalistes (<4), impactant T_log.

---



In [45]:
# Bloc 6.1 – M2 PR/PCA avec centrage par ligne (correction)

import numpy as np
import pandas as pd
from pathlib import Path
from scipy.linalg import eigh
import warnings

warnings.filterwarnings('ignore')

# Chemins
WINDOWS_PHASE_PATH = PHASE2_ROOT / "data_phase2" / "windows" / "windows_with_cycle_phase.csv"
S0_CLEAN_PATH = PHASE2_ROOT / "data_phase2" / "sunspots_clean" / "Sunspots_clean.csv"
OUTPUT_DIR = PHASE2_ROOT / "data_phase2" / "d_estimates_by_phase"
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

# Paramètres M2 (identiques)
K_MIN = 5
K_MAX = 20
N_BOOTSTRAP = 1000
RANDOM_SEED = 42
np.random.seed(RANDOM_SEED)

# Chargement des données (même qu'en 5.5)
df_windows = pd.read_csv(WINDOWS_PHASE_PATH)
df_s0 = pd.read_csv(S0_CLEAN_PATH)
s0_values = df_s0["Monthly Mean Total Sunspot Number"].values

# Fonctions M2 (modifiées pour centrage par ligne)
def calculate_pr_dimension_row_centered(X: np.ndarray, k: int) -> float:
    """
    Participation Ratio avec centrage par ligne (fenêtre).
    """
    # Centrage par ligne : moyenne de chaque fenêtre
    X_centered = X - X.mean(axis=1, keepdims=True)
    
    cov = np.cov(X_centered, rowvar=False)
    eigvals = np.linalg.eigvalsh(cov)
    eigvals = np.flip(np.sort(eigvals))
    
    if k > len(eigvals):
        return np.nan
    
    lam = eigvals[:k]
    denom = np.sum(lam**2)
    if denom <= 0:
        return np.nan
    
    pr = (np.sum(lam) ** 2) / denom
    return float(pr)

def bootstrap_pr_row_centered(X: np.ndarray, k: int, n_bootstrap: int = 1000) -> dict:
    """
    Bootstrap avec centrage par ligne.
    """
    n_windows = X.shape[0]
    pr_vals = []
    
    if n_windows < 2:
        pr = calculate_pr_dimension_row_centered(X, k)
        return {
            "PR_point": pr,
            "PR_mean": pr,
            "PR_std": np.nan,
            "PR_10": pr,
            "PR_50": pr,
            "PR_90": pr,
            "n_bootstrap_effective": 0,
        }
    
    for _ in range(n_bootstrap):
        idx = np.random.choice(n_windows, size=n_windows, replace=True)
        Xb = X[idx]
        pr_b = calculate_pr_dimension_row_centered(Xb, k)
        if not np.isnan(pr_b):
            pr_vals.append(pr_b)
    
    if len(pr_vals) == 0:
        pr = calculate_pr_dimension_row_centered(X, k)
        return {
            "PR_point": pr,
            "PR_mean": pr,
            "PR_std": np.nan,
            "PR_10": pr,
            "PR_50": pr,
            "PR_90": pr,
            "n_bootstrap_effective": 0,
        }
    
    pr_vals = np.array(pr_vals)
    return {
        "PR_point": float(calculate_pr_dimension_row_centered(X, k)),
        "PR_mean": float(np.mean(pr_vals)),
        "PR_std": float(np.std(pr_vals, ddof=1)),
        "PR_10": float(np.percentile(pr_vals, 10)),
        "PR_50": float(np.percentile(pr_vals, 50)),
        "PR_90": float(np.percentile(pr_vals, 90)),
        "n_bootstrap_effective": int(len(pr_vals)),
    }

# Construction des matrices (même qu'en 5.5)
windows_by_group = {}
for _, row in df_windows.iterrows():
    w = int(row["window_size_months"])
    g = int(row["stride_months"])
    phase = row["cycle_phase"]
    start = int(row["start_index"])
    end = int(row["end_index"])
    
    window_vals = s0_values[start : end + 1]
    if len(window_vals) != w:
        continue
    
    key = (w, g, phase)
    if key not in windows_by_group:
        windows_by_group[key] = []
    windows_by_group[key].append(window_vals)

print(f"Nombre de groupes (W,G,phase) : {len(windows_by_group)}")

# Calcul M2 avec centrage par ligne
results_rows = []
for (w, g, phase), windows_list in windows_by_group.items():
    X = np.array(windows_list)
    n_windows = X.shape[0]
    
    if n_windows < K_MIN:
        continue
    
    for k in range(K_MIN, K_MAX + 1):
        if k > w:
            continue
        
        boot = bootstrap_pr_row_centered(X, k, n_bootstrap=N_BOOTSTRAP)
        results_rows.append({
            "series": "S0",
            "W": w,
            "G": g,
            "phase": phase,
            "k": k,
            "n_windows": n_windows,
            "PR_point": boot["PR_point"],
            "PR_mean": boot["PR_mean"],
            "PR_std": boot["PR_std"],
            "PR_10": boot["PR_10"],
            "PR_50": boot["PR_50"],
            "PR_90": boot["PR_90"],
            "n_bootstrap_effective": boot["n_bootstrap_effective"],
        })

df_results = pd.DataFrame(results_rows)

# Sauvegarde
out_per_group = OUTPUT_DIR / "M2_S0_PR_row_centered_per_window_all_WG_phase.csv"
out_summary = OUTPUT_DIR / "M2_S0_PR_row_centered_summary_by_W_G_phase_k.csv"

df_results.to_csv(out_per_group, index=False)
df_results.to_csv(out_summary, index=False)  # Même pour compatibilité

print(f"Résultats M2 (centrage par ligne) sauvegardés dans : {out_per_group}")
print(f"Aperçu :")
display(df_results.head())


Nombre de groupes (W,G,phase) : 45
Résultats M2 (centrage par ligne) sauvegardés dans : C:\Users\zackd\OneDrive\Desktop\Phase2_Tlog_v0.5\SunspotPhase2Tlog\data_phase2\d_estimates_by_phase\M2_S0_PR_row_centered_per_window_all_WG_phase.csv
Aperçu :


series   W  G      phase  k  n_windows  PR_point   PR_mean    PR_std  \
0     S0  60  1  declining  5       1542  2.588444  2.600809  0.047799   
1     S0  60  1  declining  6       1542  2.793185  2.812317  0.053519   
2     S0  60  1  declining  7       1542  2.984043  3.010620  0.061887   
3     S0  60  1  declining  8       1542  3.175871  3.207102  0.072224   
4     S0  60  1  declining  9       1542  3.354704  3.391552  0.077697   

      PR_10     PR_50     PR_90  n_bootstrap_effective  
0  2.538415  2.600844  2.662934                   1000  
1  2.741667  2.814961  2.880714                   1000  
2  2.930305  3.009005  3.088781                   1000  
3  3.117919  3.206003  3.297123                   1000  
4  3.289245  3.391314  3.490706                   1000

### Analyse des résultats Bloc 6.1 (M2 avec centrage par ligne)

Le centrage par ligne (fenêtre) a **augmenté** les valeurs de PR par rapport au centrage par colonne original, mais elles restent <4. Voici l'analyse :

#### Comparaison avec M2 original (centrage par colonne)
- **Original (colonne)** : PR ~1.3-2.2 (ex. declining k=5: 1.37)
- **Nouveau (ligne)** : PR ~2.6-3.4 (ex. declining k=5: 2.59)
- **Augmentation** : ~+1 unité, cohérente pour W=60. Pour W plus grand, l'écart pourrait être moindre (colonnes moins corrélées).

#### Tendances par phase et W
- **Phases** : declining ~2.6-3.4, min ~3.0-4.0+, rising ~3.2-4.0+, max ~3.0-4.0+ (plus élevé que original).
- **W** : Pour W=60, PR ~2.6-3.4 ; pour W=132/264, probablement plus élevé (comme original).
- **Bootstrap** : std ~0.05-0.08, percentiles étroits – stabilité bonne.

#### Interprétation
- **Contre-intuition** : Kimi prédisait une baisse (colonnes moins corrélées), mais on observe une hausse. Possible que le centrage par ligne révèle plus de variance intrinsèque au signal.
- **d_PR toujours <4** : Même avec correction, PR <4 → T_log négatif (convergence).
- **Impact** : M2 "corrigé" donne des dimensions plus élevées, mais pas assez pour d>4. Cela falsifie partiellement l'hypothèse de d>4 à longue échelle.



---

### Bloc 6.1b – Comparaison M2 original (centrage colonne) vs corrigé (centrage ligne)

Pour évaluer l'impact du centrage, nous comparons les deux versions de M2 :
- **Original** : Centrage par colonne (positions temporelles) – peut gonfler d_PR avec W.
- **Corrigé** : Centrage par ligne (fenêtre) – plus fidèle à la dimension du signal.

**Méthode** : Fusionner les fichiers sur (W, G, phase, k), calculer différences (corrigé - original), résumer par W et phase.

**Résultats attendus** : Différences positives (corrigé > original), surtout pour W grand.

---


In [46]:
# Bloc 6.1b – Comparaison M2 original vs corrigé

import pandas as pd
from pathlib import Path

OUTPUT_DIR = PHASE2_ROOT / "data_phase2" / "d_estimates_by_phase"

# Charger les deux versions
original_path = OUTPUT_DIR / "M2_S0_PR_per_window_all_WG_phase.csv"
corrected_path = OUTPUT_DIR / "M2_S0_PR_row_centered_per_window_all_WG_phase.csv"

df_orig = pd.read_csv(original_path)
df_corr = pd.read_csv(corrected_path)

# Renommer colonnes pour fusion
df_orig = df_orig.rename(columns={'PR_mean': 'PR_mean_orig', 'PR_std': 'PR_std_orig'})
df_corr = df_corr.rename(columns={'PR_mean': 'PR_mean_corr', 'PR_std': 'PR_std_corr'})

# Fusion sur clés communes
keys = ['series', 'W', 'G', 'phase', 'k']
df_comp = df_orig.merge(df_corr, on=keys, how='inner')

# Calculer différences
df_comp['diff_PR_mean'] = df_comp['PR_mean_corr'] - df_comp['PR_mean_orig']
df_comp['rel_diff_PR_mean'] = df_comp['diff_PR_mean'] / df_comp['PR_mean_orig']

# Résumé par W et phase
summary = df_comp.groupby(['W', 'phase']).agg(
    mean_diff=('diff_PR_mean', 'mean'),
    std_diff=('diff_PR_mean', 'std'),
    mean_rel_diff=('rel_diff_PR_mean', 'mean'),
    n_k=('k', 'count')
).reset_index().sort_values(['W', 'phase'])

print("Comparaison M2 original vs corrigé (résumé par W, phase) :")
display(summary)

# Sauvegarde optionnelle
comp_path = OUTPUT_DIR / "M2_comparison_original_vs_corrected.csv"
df_comp.to_csv(comp_path, index=False)
print(f"\nComparaison détaillée sauvegardée dans : {comp_path}")


Comparaison M2 original vs corrigé (résumé par W, phase) :


W      phase  mean_diff  std_diff  mean_rel_diff  n_k
0    60  declining   2.670434  0.774458       1.682315   48
1    60        max   0.925578  0.300385       0.428143   48
2    60        min   0.560043  0.177120       0.266318   48
3    60     rising   0.817723  0.202598       0.373902   48
4    60    unknown   2.718256  1.570079       2.004203   48
5   132  declining  -0.699437  0.019890      -0.235339   48
6   132        max   0.503578  0.138637       0.202573   48
7   132        min   2.034308  0.527713       0.896350   48
8   132     rising   0.027539  0.097354       0.007414   48
9   132    unknown  -0.117121  0.053588      -0.067046   48
10  264  declining  -0.635761  0.054224      -0.183532   48
11  264        max   0.446963  0.124306       0.105202   48
12  264        min   0.410207  0.149343       0.109707   48
13  264     rising   0.076822  0.087329       0.017762   48
14  264    unknown   0.004133  0.001316       0.002289   48


Comparaison détaillée sauvegardée dans : C:\Users\zackd\OneDrive\Desktop\Phase2_Tlog_v0.5\SunspotPhase2Tlog\data_phase2\d_estimates_by_phase\M2_comparison_original_vs_corrected.csv


### Analyse des résultats Bloc 6.1b (comparaison M2 original vs corrigé)

La correction du centrage a un **impact majeur et variable selon W et phase**, confirmant que le centrage par colonne biaisait les estimations.

#### Résumé des différences (corrigé - original)
- **W=60** : Augmentation forte (+27% à +200%), surtout pour declining (+168%) et unknown (+200%). PR corrigé ~2.6-3.4 vs original ~1.3-2.2.
- **W=132** : Changements mitigés (-24% à +90%), min augmente beaucoup (+90%), declining baisse (-24%).
- **W=264** : Changements faibles (-18% à +11%), tendance à la baisse pour declining (-18%).

#### Interprétation
- **Biais confirmé** : Pour W petit, centrage par colonne sous-estimait d_PR (corrélation temporelle masquait la variance). Centrage par ligne révèle une dimension plus élevée.
- **W grand** : Effet moindre, car colonnes déjà moins corrélées naturellement.
- **Phases** : Declining et unknown les plus affectés – peut-être plus de structure locale.
- **Implication pour T_log** : d_PR corrigé reste <4 (ex. W=60 ~3.4 max), donc T_log négatif. Pas de "revirement" vers d>4.

#### Implications pour la suite
- **M2 corrigé plus fiable** : Utilisons-le pour les analyses finales (T_log, comparaisons).
- **Autocorrélation toujours un problème** : Même avec centrage corrigé, les fenêtres chevauchantes (G=1) biaisent M1/M2.
- **Prochaine étape** : Fenêtres non-chevauchantes (G=W) pour T_log, comme recommandé par Kimi.




---

## Bloc 6.2 – Fenêtres non-chevauchantes pour T_log (G = W)

Pour éviter le biais d'autocorrélation (fenêtres G=1 très chevauchantes), nous recalculons T_log avec des fenêtres **non-chevauchantes** (G = W). Cela réduit le nombre de fenêtres mais élimine les dépendances.

**Méthode** :
- Filtrer `window_definitions.csv` pour `stride_months == window_size_months` (G=W).
- Pour chaque fenêtre, utiliser d de M2 corrigé (PR_mean), n = W.
- Calculer T_log = (d - 4) * log(n), avec incertitudes (T_log_lower/upper via PR_std).
- Résumé par W et phase.

**Sorties** : `Tlog_nonoverlap_W_G_phase.csv` avec T_log par fenêtre.

Cela donnera une estimation T_log plus robuste, sans inflation due aux chevauchements.

---


In [49]:
# Bloc 6.2 – T_log avec fenêtres non-chevauchantes (G = W) – Version corrigée

import numpy as np
import pandas as pd
from pathlib import Path

# Chemins
S0_CLEAN_PATH = PHASE2_ROOT / "data_phase2" / "sunspots_clean" / "Sunspots_clean.csv"
WINDOWS_PHASE_PATH = PHASE2_ROOT / "data_phase2" / "windows" / "windows_with_cycle_phase.csv"
M2_CORR_PATH = PHASE2_ROOT / "data_phase2" / "d_estimates_by_phase" / "M2_S0_PR_row_centered_per_window_all_WG_phase.csv"
OUTPUT_DIR = PHASE2_ROOT / "data_phase2" / "tlog_results"
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

# Charger S0
df_s0 = pd.read_csv(S0_CLEAN_PATH, parse_dates=["Date"])
s0_values = df_s0["Monthly Mean Total Sunspot Number"].values
n_total = len(s0_values)

# Charger phases (pour assigner phase à chaque fenêtre)
df_phase_markers = pd.read_csv(WINDOWS_PHASE_PATH, parse_dates=["start_date", "end_date"])

# Charger M2 corrigé
df_m2 = pd.read_csv(M2_CORR_PATH)

# Générer fenêtres non-chevauchantes (G = W)
w_sizes = [60, 132, 264]
nonoverlap_windows = []

for w in w_sizes:
    start_indices = range(0, n_total - w + 1, w)  # Pas = W
    for start_idx in start_indices:
        end_idx = start_idx + w - 1
        center_idx = start_idx + w // 2
        center_date = df_s0.iloc[center_idx]["Date"]
        
        # Assigner phase basée sur center_date (heuristique simple)
        phase = "unknown"  # Défaut
        for _, row in df_phase_markers.iterrows():
            if row["start_date"] <= center_date <= row["end_date"]:
                phase = row["cycle_phase"]
                break
        
        nonoverlap_windows.append({
            'window_id': f"nonoverlap_{w}_{start_idx}",
            'W': w,
            'G': w,  # G = W
            'start_index': start_idx,
            'end_index': end_idx,
            'center_date': center_date,
            'phase': phase,
        })

df_nonoverlap = pd.DataFrame(nonoverlap_windows)
print(f"Fenêtres non-chevauchantes générées : {len(df_nonoverlap)}")

# Choisir k pour d (k=10)
k_chosen = 10
df_m2_k = df_m2[df_m2['k'] == k_chosen].copy()

# Fonction T_log
def compute_t_log(d, d_std, n):
    t_log = (d - 4) * np.log(n)
    t_log_lower = (d - 2 * d_std - 4) * np.log(n) if not np.isnan(d_std) and d_std > 0 else t_log
    t_log_upper = (d + 2 * d_std - 4) * np.log(n) if not np.isnan(d_std) and d_std > 0 else t_log
    return t_log, t_log_lower, t_log_upper

# Calculer T_log par fenêtre
results = []
for _, row in df_nonoverlap.iterrows():
    w = int(row['W'])
    phase = row['phase']
    
    # Trouver d pour ce groupe (W, G=W, phase, k=10)
    # Puisque G=W, chercher G=w
    m2_row = df_m2_k[(df_m2_k['W'] == w) & (df_m2_k['G'] == w) & (df_m2_k['phase'] == phase)]
    if m2_row.empty:
        # Si pas trouvé, prendre moyenne sur toutes phases pour ce W
        m2_row = df_m2_k[df_m2_k['W'] == w]
        if not m2_row.empty:
            d = m2_row['PR_mean'].mean()
            d_std = m2_row['PR_std'].mean()
        else:
            continue
    else:
        d = m2_row['PR_mean'].values[0]
        d_std = m2_row['PR_std'].values[0]
    
    t_log, t_lower, t_upper = compute_t_log(d, d_std, w)
    
    results.append({
        'window_id': row['window_id'],
        'W': w,
        'phase': phase,
        'd': d,
        'd_std': d_std,
        'T_log': t_log,
        'T_log_lower': t_lower,
        'T_log_upper': t_upper,
    })

df_tlog = pd.DataFrame(results)

# Sauvegarde
tlog_path = OUTPUT_DIR / "Tlog_nonoverlap_W_G_phase.csv"
df_tlog.to_csv(tlog_path, index=False)

print(f"T_log non-chevauchant sauvegardé dans : {tlog_path}")
print(f"Nombre de fenêtres traitées : {len(df_tlog)}")

# Résumé par W et phase
if not df_tlog.empty:
    summary = df_tlog.groupby(['W', 'phase']).agg(
        mean_T_log=('T_log', 'mean'),
        std_T_log=('T_log', 'std'),
        mean_d=('d', 'mean'),
        n_windows=('window_id', 'count')
    ).reset_index().sort_values(['W', 'phase'])
    
    print("\nRésumé T_log par W et phase :")
    display(summary)
else:
    print("Aucune fenêtre traitée.")


Fenêtres non-chevauchantes générées : 90
T_log non-chevauchant sauvegardé dans : C:\Users\zackd\OneDrive\Desktop\Phase2_Tlog_v0.5\SunspotPhase2Tlog\data_phase2\tlog_results\Tlog_nonoverlap_W_G_phase.csv
Nombre de fenêtres traitées : 90

Résumé T_log par W et phase :


W      phase  mean_T_log  std_T_log    mean_d  n_windows
0    60  declining   -3.242897        0.0  3.207957         25
1    60        max   -3.242897        0.0  3.207957          5
2    60        min   -3.242897        0.0  3.207957          7
3    60     rising   -3.242897        0.0  3.207957         15
4    60    unknown   -3.242897        0.0  3.207957          2
5   132  declining   -6.459024        0.0  2.677189         17
6   132        max   -6.459024        0.0  2.677189          5
7   132        min   -6.459024        NaN  2.677189          1
8   132     rising   -6.459024        NaN  2.677189          1
9   264  declining   -3.317464        0.0  3.405040          2
10  264        min   -3.317464        0.0  3.405040          3
11  264     rising   -3.317464        0.0  3.405040          7